process/block/preprocess/transactions.go

```diff
@@ -1317,8 +1343,41 @@ func (txs *transactions) IsInterfaceNil() bool {
 	return txs == nil
 }
 
-// SortTransactionsBySenderAndNonce sorts the provided transactions and hashes simultaneously
-func SortTransactionsBySenderAndNonce(transactions []*txcache.WrappedTransaction) {
+// sortTransactionsBySenderAndNonce sorts the provided transactions and hashes simultaneously
+func (txs *transactions) sortTransactionsBySenderAndNonce(transactions []*txcache.WrappedTransaction, randomness []byte) {
+	if !txs.flagFrontRunningProtection.IsSet() {
+		sortTransactionsBySenderAndNonceLegacy(transactions)
+		return
+	}
+
+	txs.sortTransactionsBySenderAndNonceWithFrontRunningProtection(transactions, randomness)
+}
+
+func (txs *transactions) sortTransactionsBySenderAndNonceWithFrontRunningProtection(transactions []*txcache.WrappedTransaction, randomness []byte) {
+	// make sure randomness is 32bytes and uniform
+	randSeed := txs.hasher.Compute(string(randomness))
+	xoredAddresses := make(map[string][]byte)
+
+	for _, tx := range transactions {
```

Comment:
can we add the sorting to be done By GasPrice buckets ? so sort those with gasPrice X - after that with GasPrice with X-1 and so on ? or buckets in GasPrice of intervals of 100 ?

Reply:
I think this will defeat the purpose of front running protection. The higher the gas price on a transaction, the higher chances to be selected on every block to executed. Indeed, it might not get to be executed in the first block after entering the pool but will eventually be executed as the randomness will yield at some point a favorable value.

Reply:
currently anyway we prefilter the "first batch" of transactions, which we got from the smart tx cache, fitting the allowed "gas bandwidth". The smart txcache gives you the higher scored senders first, so in a way there is a priority to get into the batch for the higher price transactions, but the order inside the batch is not guaranteed, as it depends on the randomness.
This is I think the effect we wanted.